src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -11848,6 +11904,99 @@ class CL1<T>
                 );
         }
 
+        [Fact]
+        public void Lambda_18()
+        {
+            var source =
+@"delegate T D<T>(T t) where T : class;
+class C
+{
+    static void F()
+    {
+        var d1 = (D<string?>)((string s1) =>
+            {
+                s1 = null;
+                return s1;
+            });
+        var d2 = (D<string>)((string? s2) =>
+            {
+                s2.ToString();
+                return s2;
+            });
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (8,22): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //                 s1 = null;
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "null").WithLocation(8, 22),
+                // (13,17): warning CS8602: Possible dereference of a null reference.
+                //                 s2.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s2").WithLocation(13, 17),
+                // (14,24): warning CS8603: Possible null reference return.
+                //                 return s2;
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "s2").WithLocation(14, 24));
+        }
+
+        /// <summary>
+        /// Inferred nullability of captured variables outside lambda
+        /// should be independent of inferred nullability inside lambda.
+        /// </summary>
+        [Fact]
+        public void Lambda_19()
+        {
+            var source =
+@"using System;
+class C
+{
+    static void F1(object? x1, object y1)
+    {
+        object z1 = y1;
+        Action f = () =>
+        {
+            z1 = x1; // warning
+        };
+        f();
+        z1.ToString();
```

Comment:
>z1 [](start = 8, length = 2)

It looks to me like there should be a diagnostic here. As I understand the LDM's decisions, we should be treating the lambda as if it might be executed where it appears. Since it might be executed here, it might leave a null in z1.

Reply:
Yes there should be a diagnostic here. This PR was created before the decision on how to track nullability of captured variables inside and outside the lambda, so this PR uses the simplest approach of treating the nullability of variables inside and outside the lambda as independent.

I'll add `// PROTOTYPE` comments to `NullableWalker.GetVariableState` and to this test and other tests that should be updated.

---
In reply to: [193917896](https://github.com/dotnet/roslyn/pull/27283#discussion_r193917896) [](ancestors = 193917896)